File: cs/Refactor1/SRP_DI_Workshop/Domain/Revenue/PriceCalculator.cs
using System;

namespace SRP_DI_Workshop.Domain.Revenue
{
    public class PriceCalculator
    {
        public void CalculatePrice(OrderItem item, InventoryItem inventoryItem)
        {
            IPricingCalculation pricingCalculation;

            switch (inventoryItem.PricingCalculation)
            {
                case FullQuantityPricingCalculation.Id:
                    pricingCalculation = new FullQuantityPricingCalculation();
                    break;
                case Buy2Get1FreePricingCalculation.Id:
                    pricingCalculation = new Buy2Get1FreePricingCalculation();
                    break;
                default:
                    throw new InvalidOperationException("Invalid pricing calculation type: " + inventoryItem.PricingCalculation);
            }

            pricingCalculation.CalculatePrice(item, inventoryItem);
        }
    }
}

File: java/Refactor2/src/main/java/no/bouvet/solid/srpdip/domain/revenue/PriceCalculator.java
package no.bouvet.solid.srpdip.domain.revenue;

import no.bouvet.solid.srpdip.Factory;
import no.bouvet.solid.srpdip.domain.InventoryItem;
import no.bouvet.solid.srpdip.domain.OrderItem;

public class PriceCalculator
{
	public static Factory<PriceCalculator> factory = new Factory<>(PriceCalculator.class);
	
    public void calculate(OrderItem item, InventoryItem inventoryItem)
    {
        PricingCalculation pricingCalculation;

        switch (inventoryItem.getPricingCalculation())
        {
            case "FULLQUANTITY":
                pricingCalculation = FullQuantityPricingCalculation.factory.getInstance();
                break;
            case "BUY2GET1FREE":
                pricingCalculation = Buy2Get1FreePricingCalculation.factory.getInstance();
                break;
            default:
                throw new RuntimeException("Invalid pricing calculation type: " + inventoryItem.getPricingCalculation());
        }

        pricingCalculation.calculate(item, inventoryItem);
    }
}
